### PR TITLE
Network switching #1091

### PR DESCRIPTION
**Pull request by rasom**
Should be merged after https://github.com/status-im/status-react/pull/1584

# What is left on status-react side
- new tests for `status-im.ui.screens.network-settings.events` and `status-im.ui.screens.accounts.login.events`
- [x] migrations for accounts (default netwroks and network should be added to accounts)
- [x] just found that I missed icons from previous PR https://github.com/status-im/status-react/pull/1115

# What is left on status-go side 

- [x] https://github.com/status-im/status-go/pull/272
- [x] https://github.com/status-im/status-go/issues/275
- [x] https://github.com/status-im/status-go/issues/274
- [x] https://github.com/status-im/status-go/issues/282 
- [x] https://github.com/status-im/status-go/issues/251

Also in case if we change status-go version in this PR (and we will), we have to change it here as well https://github.com/status-im/react-native-webview-bridge/blob/master/android/build.gradle#L34

status: ready

### CODE REVIEW

**Comment by flexsurfer** [comment]
status-im/status-go#282 (Blocker) probably on RN side `Logged in account: {"error":"cannot retrieve a valid key for a given account: no key for given address or file"}`

File: modules/react-native-status/ios/RCTStatus/RCTStatus.m [hunk 1]
```diff
@@ -101,157 +101,101 @@ -(void)setBridge:(RCTBridge *)newBridge
...
+    NSLog(@"logUrlPath %@", logUrl.path);
+    if(![fileManager fileExistsAtPath:logUrl.path]) {
         NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
         [dict setObject:[NSNumber numberWithInt:511] forKey:NSFilePosixPermissions];
-        [manager createFileAtPath:logUrl.path contents:nil attributes:dict];
+        [fileManager createFileAtPath:logUrl.path contents:nil attributes:dict];
+    }
 #ifndef DEBUG
-        [Instabug addFileAttachmentWithURL:[folderName URLByAppendingPathComponent:@"geth.log"]];
+    //[Instabug addFileAttachmentWithURL:logUrl];
```
Comment (oskarth) [inline modules/react-native-status/ios/RCTStatus/RCTStatus.m]: What does this have to do with network switching?

File: modules/react-native-status/ios/RCTStatus/RCTStatus.m [hunk 2]
```diff
@@ -101,157 +101,101 @@ -(void)setBridge:(RCTBridge *)newBridge
...
+    NSError *error = nil;
+    NSURL *rootUrl =[[fileManager
+                      URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask]
+                     lastObject];
+    NSURL *testnetFolderName = [rootUrl URLByAppendingPathComponent:@"ethereum/testnet"];
+    
+    if (![fileManager fileExistsAtPath:testnetFolderName.path])
+        [fileManager createDirectoryAtPath:testnetFolderName.path withIntermediateDirectories:YES attributes:nil error:&error];
+    
+    NSURL *flagFolderUrl = [rootUrl URLByAppendingPathComponent:@"ropsten_flag"];
```
Comment (oskarth) [inline modules/react-native-status/ios/RCTStatus/RCTStatus.m:121]: Does this mean ropsten has special role? Is it a matter of it being default somehow?

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java [hunk 3]
```diff
@@ -322,6 +357,20 @@ public void run() {
     }
 
     @ReactMethod
+    public void stopNode() {
+
+        Thread thread = new Thread() {
+            @Override
+            public void run() {
+                Log.d(TAG, "stopNode");
```
Comment (oskarth) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:365]: StopNode if we want to be consistent

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java [hunk 4]
```diff
@@ -167,7 +204,7 @@ private void doStartNode() {
             Uri gethLogUri = Uri.fromFile(logFile);
             try {
                 Log.d(TAG, "Attach to geth.log to instabug " + gethLogUri.getPath());
-                Instabug.setFileAttachment(gethLogUri, gethLogFileName);
+                //Instabug.setFileAttachment(gethLogUri, gethLogFileName);
```
Comment (oskarth) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java]: Same here, relationship to network switching?

File: .env.jenkins [hunk 5]
```diff
@@ -3,4 +3,4 @@ WALLET_WIP_ENABLED=1
 NOTIFICATIONS_WIP_ENABLED=1
 DEBUG_LOGS_ENABLED=1
 STUB_STATUS_GO=0
-
+NETWORK_SWITCHING=1
```
Comment (oskarth) [inline .env.jenkins:6]: Can we add explicit NETWORK_SWITCHING=0 to prod?

File: src/status_im/components/drawer/view.cljs [hunk 6]
```diff
@@ -145,10 +145,11 @@
...
 
-(defn current-network []
-  [view {:style st/network-label-container}
-   [text {:style st/network-label} (i18n/label :t/current-network)]
-   [text {:style st/network-title} "Ropsten"]])
+(defview current-network []
+  (letsubs [{:keys [name]} [:get-current-account-network]]
+    [view {:style st/network-label-container}
+     [text {:style st/network-label} (i18n/label :t/current-network)]
+     [text {:style st/network-title} name]]))
```
Comment (oskarth) [inline src/status_im/components/drawer/view.cljs]: Do we want to shadow `name`? It's a builtin. Can we do this with `(:name m)` instead?

File: src/status_im/components/icons/vector_icons.cljs [hunk 7]
```diff
@@ -64,7 +64,7 @@
             :icons/dropdown-up         (slurp-svg "./resources/icons/dropdown_up.svg")
             :icons/dropdown            (slurp-svg "./resources/icons/dropdown.svg")
             :icons/grab                (slurp-svg "./resources/icons/grab.svg")
-            :icons/share               (slurp-svg "./resources/icons/share.svg")})
```
Comment (oskarth) [inline src/status_im/components/icons/vector_icons.cljs:67]: What's share.svg? Can we remove this file?

File: src/status_im/data_store/accounts.cljs [hunk 8]
```diff
@@ -9,6 +9,3 @@
 
 (defn save [account update?]
   (data-store/save account update?))
-
-(defn save-all [accounts update?]
-  (data-store/save-all accounts update?))
```
Comment (oskarth) [inline src/status_im/data_store/accounts.cljs:14]: Why is this gone with network switching?

File: src/status_im/data_store/networks.cljs [hunk 9]
```diff
@@ -0,0 +1,23 @@
+(ns status-im.data-store.networks
+  (:require [status-im.data-store.realm.networks :as data-store])
+  (:refer-clojure :exclude [exists?]))
```
Comment (oskarth) [inline src/status_im/data_store/networks.cljs]: Seems like exists? is namespaced, is this necessary then?

File: src/status_im/data_store/networks.cljs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  [{:keys [id] :as network}]
+  (data-store/save network (data-store/exists? id)))
+
+(defn save-all
+  [networks]
+  (mapv save networks))
+
+(defn save-property
+  [id property-name value]
+  (data-store/save-property id property-name value))
```
Comment (oskarth) [inline src/status_im/data_store/networks.cljs:18]: What's a property?

File: modules/react-native-status/ios/RCTStatus/RCTStatus.m (hunk 2 above)
Comment (rasom) [reply modules/react-native-status/ios/RCTStatus/RCTStatus.m:121]: This is related to migration to ropsten network from previous testnet (don't remember the name), you can find this code in develop as well.

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 3 above)
Comment (rasom) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:365]: ok

File: src/status_im/data_store/realm/accounts.cljs [hunk 11]
```diff
@@ -1,20 +1,25 @@
...
 
 (defn save [account update?]
   (realm/write realm/base-realm
-    #(realm/create realm/base-realm :account account update?)))
-
-(defn save-all [accounts update?]
-  (realm/write realm/base-realm
-               (fn []
-                 (mapv #(realm/create realm/base-realm :account % update?) accounts))))
+    (let [account' (update account :networks (comp vec vals))]
```
Comment (oskarth) [inline src/status_im/data_store/realm/accounts.cljs:24]: Can we call this something other than `account'`? Why do we get rid of network keys (`vals`) when persisting?

File: src/status_im/components/icons/vector_icons.cljs (hunk 7 above)
Comment (flexsurfer) [reply src/status_im/components/icons/vector_icons.cljs:67]: i think it happened during the conflict, we should keep both `share` and `network` icons

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 4 above)
Comment (rasom) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java]: probably not, it's related to some intermediate testing

File: .env.jenkins (hunk 5 above)
Comment (rasom) [reply .env.jenkins:6]: ok

File: src/status_im/data_store/realm/schemas/base/v4/account.cljs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+                                                :objectType :network}}})
+
+(defn migration [_old-realm new-realm]
+  (log/debug "migrating account schema v4")
+  (let [accounts (.objects new-realm "account")]
+    (dotimes [i (.-length accounts)]
+      (let [account (aget accounts i)]
+        (aset account "network" constants/default-network)))))
+        ;; creation of network object crashes the app
+        ;(aset account "networks" (clj->js (vals constants/default-networks)))))))
```
Comment (oskarth) [inline src/status_im/data_store/realm/schemas/base/v4/account.cljs]: What's this comment? Don't understand

File: src/status_im/data_store/realm/schemas/base/v4/network.cljs [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+             :properties {:id      :string
+                          :name    {:type     :string
+                                    :optional true}
+                          :config  {:type     :string
+                                    :optional true}
+                          :rpc-url {:type     :string
+                                    :optional true}}})
+
+(defn migration [old-realm new-realm]
+  (log/debug "migrating group schema v4"))
```
Comment (oskarth) [inline src/status_im/data_store/realm/schemas/base/v4/network.cljs]: Is this group? Not network?

File: src/status_im/components/icons/vector_icons.cljs (hunk 7 above)
Comment (rasom) [reply src/status_im/components/icons/vector_icons.cljs:67]: Don't think so, that's probably result of wrong conflict resolving, will fix

File: src/status_im/native_module/impl/non_status_go_module.cljs [hunk 14]
```diff
@@ -15,6 +15,7 @@
   (-start-node [this callback]
```
Comment (oskarth) [inline src/status_im/native_module/impl/non_status_go_module.cljs]: config

File: src/status_im/data_store/realm/schemas/base/v4/account.cljs (hunk 12 above)
Comment (rasom) [reply src/status_im/data_store/realm/schemas/base/v4/account.cljs]: @flexsurfer please elaborate

File: src/status_im/translations/en.cljs [hunk 15]
```diff
@@ -373,4 +373,24 @@
    :transactions-filter-title             "Filter History"
    :transactions-filter-tokens            "Tokens"
    :transactions-filter-type              "Type"
-   :transactions-filter-select-all        "Select all"})
+   :transactions-filter-select-all        "Select all"
+
+   ;network settings
+   :new-network                           "New network"
+   :add-network                           "Add network"
+   :add-new-network                       "Add new network"
```
Comment (oskarth) [inline src/status_im/translations/en.cljs:381]: Haven't seen UI, but why do we have 3 that seem to kind of do same thing?

File: src/status_im/ui/screens/accounts/events.cljs [hunk 16]
```diff
@@ -96,11 +96,16 @@
...
   :add-account
-  (fn [{{:keys [network] :as db} :db} [_ {:keys [address] :as account} password]]
-    (let [account' (assoc account :network network)]
-      {:db             (assoc-in db [:accounts/accounts address] account')
-       ::save-account  account'
-       :dispatch-later [{:ms 400 :dispatch [:login-account address password true]}]})))
+  (fn [{{:keys          [network]
+         :networks/keys [networks]
+         :as            db} :db} [_ {:keys [address] :as account} password]]
+    (let [account' (assoc account :network network
```
Comment (oskarth) [inline src/status_im/ui/screens/accounts/events.cljs:102]: account-with-networks maybe?

File: src/status_im/ui/screens/accounts/login/events.cljs [hunk 17]
```diff
@@ -40,27 +43,73 @@
...
   :login-account
-  (fn [{db :db} [_ address password account-creation?]]
-    {:db     (-> db
-                 (assoc :accounts/account-creation? account-creation?)
-                 (assoc-in [:accounts/login :processing] true))
-     ::login [address password]}))
+  (fn [{{:keys [network status-node-started?] :as db} :db}
+       [_ address password account-creation?]]
+    (let [{account-network :network} (get-network-by-address db address)
+          db' (-> db
```
Comment (oskarth) [inline src/status_im/ui/screens/accounts/login/events.cljs:86]: Seems to do a lot, put in separate fn and another name than db'?

File: src/status_im/ui/screens/events.cljs [hunk 18]
```diff
@@ -197,8 +196,14 @@
 
 (register-handler-fx
   :initialize-geth
-  (fn [_ _]
-    {::initialize-geth-fx nil}))
+  (fn [{db :db} _]
+    (let [{:accounts/keys [current-account-id accounts]} db
+          default-networks (get db :networks/networks)
```
Comment (oskarth) [inline src/status_im/ui/screens/events.cljs]: Why `(get db k)` instead of `(k db)`?

File: src/status_im/ui/screens/network_settings/events.cljs [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+  (fn [networks]
+    (networks/save-all networks)))
+
+;; handlers
+
+(handlers/register-handler-fx
+  :add-networks
+  (fn [{{:networks/keys [networks] :as db} :db} [_ new-networks]]
+    (let [identities    (set (keys networks))
+          new-networks' (->> new-networks
```
Comment (oskarth) [inline src/status_im/ui/screens/network_settings/events.cljs:20]: Put in separate fn? Seems to do some kind of clean up but not immediately clear what

File: src/status_im/ui/screens/network_settings/network_details/views.cljs [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+           [react/view st/connect-button
+            [react/text {:style      st/connect-button-label
+                         :uppercase? (get-in platform/platform-specific [:uppercase?])}
+             (i18n/label :t/connect)]]
+           [react/text {:style st/connect-button-description}
+            (i18n/label :t/connecting-requires-login)]]])
+       [react/view st/network-config-container
+        [react/text {:style st/network-config-text}
+         config]]
+       [react/view {:opacity 0.4}
```
Comment (oskarth) [inline src/status_im/ui/screens/network_settings/network_details/views.cljs:43]: Put in styles?

File: src/status_im/ui/screens/network_settings/parse_json/views.cljs [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+    [clojure.string :as str]))
+
+(defn paste-json-text []
+  (let [network-json "test"
+        error        nil]
+    [view {:flex 1}
+     [status-bar]
+     [toolbar {:title (i18n/label :t/add-network)}]
+     [network-settings/network-badge]
+     [view {:margin-top  16
```
Comment (oskarth) [inline src/status_im/ui/screens/network_settings/parse_json/views.cljs:20]: Styles?

File: src/status_im/ui/screens/network_settings/views.cljs [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+   [react/view {:padding-left 16}
+    [react/text {:style st/badge-name-text}
+     (or name (i18n/label :t/new-network))]
+    (when connected?
+      [react/text {:style st/badge-connected-text}
+       (i18n/label :t/connected)])]])
+
+(defn actions-view []
+  [react/view action-button-styles/actions-list
+   [react/view {:opacity 0.4}
```
Comment (oskarth) [inline src/status_im/ui/screens/network_settings/views.cljs:34]: Same as above styles, similar for other parts of this ns, padding etc

File: src/status_im/ui/screens/views.cljs [hunk 23]
```diff
@@ -91,8 +95,13 @@
                           :accounts accounts
                           :login login
                           :recover recover
+                          :network-settings network-settings
+                          :paste-json-text paste-json-text
+                          :add-rpc-url add-rpc-url
+                          :network-details network-details
```
Comment (oskarth) [inline src/status_im/ui/screens/views.cljs:101]: Value align

File: src/status_im/data_store/realm/schemas/base/v4/account.cljs (hunk 12 above)
Comment (flexsurfer) [reply src/status_im/data_store/realm/schemas/base/v4/account.cljs]: originally i implemented this using this commented line, but after updating realm, app started crashes, we had to find a workaround, i decided not to remove this, maybe for the future

File: src/status_im/utils/transactions.cljs [hunk 24]
```diff
@@ -6,6 +6,7 @@
 (defn get-network-subdomain [network]
   (case network
     "testnet" "ropsten"
+    "testnet_rpc" "ropsten"
```
Comment (oskarth) [inline src/status_im/utils/transactions.cljs:9]: What is difference in testnet vs testnet_rpc?

File: src/status_im/data_store/realm/schemas/base/v4/network.cljs (hunk 13 above)
Comment (flexsurfer) [reply src/status_im/data_store/realm/schemas/base/v4/network.cljs]: copy/paste typo

File: modules/react-native-status/ios/RCTStatus/RCTStatus.m (hunk 2 above)
Comment (oskarth) [reply modules/react-native-status/ios/RCTStatus/RCTStatus.m]: ok

File: src/status_im/components/drawer/view.cljs (hunk 6 above)
Comment (janherich) [reply src/status_im/components/drawer/view.cljs]: +1

File: src/status_im/constants.cljs [hunk 25]
```diff
@@ -26,3 +27,28 @@
 
 (def console-chat-id "console")
 (def wallet-chat-id "wallet")
+
+(def default-network "testnet_rpc")
+(def default-networks
+  {"mainnet"     {:id     "mainnet",
```
Comment (flexsurfer) [inline src/status_im/constants.cljs:33]: do we really want to give access to mainnet in this release? @sla-shi ?

Comment (sla-shi) [reply src/status_im/constants.cljs:33]: mainnet should be disabled until we tested switching between testnets @flexsurfer

Comment (flexsurfer) [reply src/status_im/constants.cljs:33]: so we need to implement this in UI then @rasom , disable mainnet in the list

File: src/status_im/data_store/realm/schemas/base/v4/account.cljs (hunk 12 above)
Comment (janherich) [reply src/status_im/data_store/realm/schemas/base/v4/account.cljs]: I think it's quite confusing, because now, when `constants/default-networks` already contains js object values, uncommenting the `clj->js` would crash it for sure.
I'm in favor of removing the line altogether, maybe adding `TODO` comment in the constants namespace.

File: src/status_im/ui/screens/accounts/events.cljs [hunk 26]
```diff
@@ -96,11 +96,16 @@
...
+                                  :networks networks)]
+      (merge
+        {:db            (assoc-in db [:accounts/accounts address] account')
+         ::save-account account'}
+        (when password
+          {:dispatch-later [{:ms 400 :dispatch [:login-account address password true]}]})))))
```
Comment (janherich) [inline src/status_im/ui/screens/accounts/events.cljs:108]: What's the reason for this delayed dispatch ?
And why do we actually use dispatch at all, instead of merging effects of `:login-account` handler into `:add-account` effects ?

File: src/status_im/ui/screens/accounts/events.cljs [hunk 27]
```diff
@@ -123,8 +128,21 @@
...
                                [address account]))
-                        (into {}))]
-      {:db (assoc db :accounts/accounts accounts)})))
+                        (into {}))
+          ;;workaround for realm bug, migrating account v4
+          events (mapv #(when (empty? (:networks %)) [:account-update-networks (:address %)]) (vals accounts))]
+      (merge
+        {:db (assoc db :accounts/accounts accounts)}
+        (when-not (empty? events)
+          {:dispatch-n events})))))
```
Comment (janherich) [inline src/status_im/ui/screens/accounts/events.cljs:137]: Again, unnecessary dispatching events.

File: src/status_im/ui/screens/accounts/login/events.cljs (hunk 17 above)
Comment (janherich) [reply src/status_im/ui/screens/accounts/login/events.cljs:86]: +1

File: src/status_im/ui/screens/events.cljs (hunk 18 above)
Comment (janherich) [reply src/status_im/ui/screens/events.cljs]: +1

File: src/status_im/ui/screens/network_settings/events.cljs [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+                             (into {}))]
+      {:db            (-> db
+                          (update :networks merge new-networks')
+                          (assoc :new-networks (vals new-networks')))
+       :save-networks new-networks'})))
+
+(handlers/register-handler-fx
+  :connect-network
+  (fn [_ [_ network]]
+    {:dispatch-n [[:account-update {:network network}]
```
Comment (janherich) [inline src/status_im/ui/screens/network_settings/events.cljs:32]: Unnecessary dispatch of `:account-update`

**Review by janherich — CHANGES_REQUESTED** [review]

File: src/status_im/ui/screens/accounts/events.cljs (hunk 26 above)
Comment (rasom) [reply src/status_im/ui/screens/accounts/events.cljs:108]: https://github.com/status-im/status-react/pull/1756#pullrequestreview-61658341

File: src/status_im/utils/transactions.cljs (hunk 24 above)
Comment (rasom) [reply src/status_im/utils/transactions.cljs:9]: second uses RPC proxy

File: src/status_im/ui/screens/views.cljs (hunk 23 above)
Comment (rasom) [reply src/status_im/ui/screens/views.cljs:101]: That doesn't work and has never worked for `case` form. I'm not going to add this manually.

File: src/status_im/ui/screens/accounts/events.cljs (hunk 16 above)
Comment (rasom) [reply src/status_im/ui/screens/accounts/events.cljs:102]: i doubt it worth efforts

File: src/status_im/data_store/realm/accounts.cljs (hunk 11 above)
Comment (rasom) [reply src/status_im/data_store/realm/accounts.cljs:24]: > Can we call this something other than account'?

why?

> Why do we get rid of network keys (vals) when persisting?

realm doesn't support this

File: src/status_im/data_store/networks.cljs (hunk 10 above)
Comment (rasom) [reply src/status_im/data_store/networks.cljs:18]: a property of an entity, no? I suppose that this boilerplate code was copied from another ns

File: src/status_im/data_store/networks.cljs (hunk 9 above)
Comment (rasom) [reply src/status_im/data_store/networks.cljs]: yeah, true. That's a legacy, don't forget this PR is on its 4th month now :D

File: src/status_im/data_store/networks.cljs (hunk 10 above)
Comment (oskarth) [reply src/status_im/data_store/networks.cljs:18]: Oh ok, I thought this was a new semantic thing related to accounts/network

File: src/status_im/data_store/realm/accounts.cljs (hunk 11 above)
Comment (oskarth) [reply src/status_im/data_store/realm/accounts.cljs:24]: Wrote in Slack but putting it here too. All these `account'` etc comments: idea is just that adding `'` is not very descriptive of what it is or how it is different, if possible it's nice to capture this in variable name imo. If not then OK

Comment (rasom) [reply src/status_im/data_store/realm/accounts.cljs:24]: code after that name can be descriptive enough. Again, imo, adding something
```clojure
(let [this-name-means-that-something-was-done-with-account (doing something with account)]
...)
```
to these intermediate symbols doesn't worth efforts. They are useful because can make code which uses them later more clear (instead of inlining some additional stuff), but eh, adding `'` is enough imo. We can move this discussion to slack, sure, i haven't seen it

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 3 above)
Comment (rasom) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:365]: Actually no, we are consistent here because that's the name of Java method.

Comment (oskarth) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java]: ok

File: src/status_im/data_store/accounts.cljs (hunk 8 above)
Comment (rasom) [reply src/status_im/data_store/accounts.cljs:14]: `src/status_im/data_store/realm/accounts.cljs` ns was touched by PR, and `save-all` from `src/status_im/data_store/realm/accounts.cljs`was used only in one place, here. So i made a small cleanup in ns that was related to PR and this affected another ns which is acceptable imo. In general i agree that touching code that is completely unrelated to PR is not desirable.

File: src/status_im/ui/screens/accounts/events.cljs (hunk 26 above)
Comment (janherich) [reply src/status_im/ui/screens/accounts/events.cljs:108]: I looked at the commentary, but I still think what we are doing there (doing some "that should be enough" delay to solve race conditions) is fundamentally wrong. 
When the problem is that effect (adding realm account and switching to it in this case) has to be already performed in order for next action (login process for the new account) to be valid, right solution is to attach callback (or better `:next-event`) to `::save-account` effect and only continue with process once effect takes place.

Comment (rasom) [reply src/status_im/ui/screens/accounts/events.cljs:108]: Yep, ok, let's file PR with this change then.

**Comment by rasom** [comment]
Other changes will be done in separate PRs.

File: src/status_im/ui/screens/accounts/events.cljs (hunk 26 above)
Comment (janherich) [reply src/status_im/ui/screens/accounts/events.cljs:108]: Correct, I will be happy to do it.

**Review by oskarth — APPROVED** [review]

**Review by flexsurfer — APPROVED** [review]

**Review by janherich — APPROVED** [review]
